Add `Agent.to_mcp()` method

Kludex opened the pull request.

Kludex committed:
Add `Agent.to_mcp()` method

Kludex commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -121,3 +129,45 @@ def map_from_sampling_content(
+    deps: AgentDepsT = None,
+) -> Server:
+    server_name = to_snake((server_name or agent.name or 'PydanticAI Agent').replace(' ', '_'))
+    tool_name = to_snake((tool_name or agent.name or 'PydanticAI Agent').replace(' ', '_'))
```
cool

Kludex commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -121,3 +129,45 @@ def map_from_sampling_content(
+            raise ValueError(f'Unknown tool: {name}')
+
+        # TODO(Marcelo): Should we pass the `message_history` instead?
+        prompt = cast(str, args['prompt'])
```
No, that would be incorrect... What I actually need to check if `'prompt'` is in `args`, and check that it's a string.

Kludex commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -121,3 +129,45 @@ def map_from_sampling_content(
+        if name != tool_name:
+            raise ValueError(f'Unknown tool: {name}')
+
+        # TODO(Marcelo): Should we pass the `message_history` instead?
```
Hmmm, I think the point is that we need to maintain the history in the session...

Good point!

DouweM replied:
Are you sure the tool should be stateful like that? If it's essentially a subagent, wouldn't multiple calls be expected to start separate subagents? I think continuing the conversation should be explicit, with some conversation ID returned and passed in

Kludex commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -121,3 +129,45 @@ def map_from_sampling_content(
+
+        return dict(result=result.output)
+
+    app.list_tools()(list_tools)
```
Decorators inside a function are treated as misused type-wise.

DouweM commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -121,3 +129,45 @@ def map_from_sampling_content(
+                name=tool_name,
+                description=tool_description,
+                inputSchema={'type': 'object', 'properties': {'prompt': {'type': 'string'}}},
+                # TODO(Marcelo): How do I get this?
```
Yep

DouweM replied:
Created an issue: https://github.com/pydantic/pydantic-ai/issues/3225

Kludex replied:
Done — `output_json_schema` exists on `AbstractAgent` since #3225 was merged, and is now used here.

Kludex committed:
Merge remote-tracking branch 'origin/main' into add-to-mcp

# Conflicts:
#	pydantic_ai_slim/pydantic_ai/agent/abstract.py

Kludex committed:
Address PR review comments for `agent_to_mcp`

- Use `output_json_schema()` for the `outputSchema` field, wrapped to match structured content shape
- Replace `cast` with proper `isinstance` validation for the `prompt` argument
- Use logfire placeholder syntax instead of f-string
- Remove resolved TODO comments
- Add tests covering tool listing, tool calling, error handling, custom names, and structured output schemas

Kludex committed:
Remove stale `pragma: no cover` on covered line

Kludex committed:
Require `prompt` in inputSchema, add docstring to `to_mcp`

- Add `'required': ['prompt']` to the inputSchema so MCP validates the field
- Add docstring to `to_mcp` matching the `to_a2a` pattern
- Update test expectations for MCP-level validation of missing prompt

Kludex committed:
Merge branch 'main' into add-to-mcp

DouweM commented:
@Kludex I think this is still interesting if you wanna get it green again

Kludex committed:
Address PR feedback: preserve user names, JSON-serialize output, fix circular import

Kludex commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -135,3 +142,49 @@ def map_from_sampling_content(
+    deps: AgentDepsT = None,
+) -> Server:
+    server_name = to_snake((server_name or agent.name or 'PydanticAI Agent').replace(' ', '_'))
+    tool_name = to_snake((tool_name or agent.name or 'PydanticAI Agent').replace(' ', '_'))
```
Fixed - `to_snake` is now only applied to the default name; user-provided `tool_name`/`server_name` are passed through unchanged.

Kludex commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -135,3 +142,53 @@ def map_from_sampling_content(
+        logfire.info('Calling tool {name}', name=name, args=args)
+
+        result = await agent.run(user_prompt=prompt, deps=deps)
+
```
Fixed - the output is now serialized via `TypeAdapter(agent.output_type).dump_python(..., mode="json")` before being returned, and a `call_tool` test for `BaseModel` output was added.

Kludex commented:
@DouweM picked this back up. Changes since last review:

- Use the new `AbstractAgent.output_json_schema()` from #3225.
- Apply `to_snake` only to the default name (derived from `agent.name`); user-provided `server_name`/`tool_name` pass through unchanged.
- JSON-serialize `result.output` via `TypeAdapter` before returning, so structured outputs (Pydantic models, dataclasses, etc.) round-trip correctly through MCP `StructuredContent`. Added a `call_tool` test exercising a `BaseModel` output.
- Drop the redundant `isinstance(prompt, str)` check - the MCP server validates `inputSchema` before the handler runs - and switch `call_tool`'s `args` to a `TypedDict`.
- Move `AbstractAgent` to `TYPE_CHECKING` to avoid a circular import via `pydantic_ai.mcp`.

Kludex committed:
Skip doctest of to_mcp example to avoid experimental warning

Kludex committed:
Restore pragma: no cover on unreachable AgentRunError

Kludex committed:
Handle OutputSpec wrappers in to_mcp; use !r in error message

Kludex commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -135,3 +145,56 @@ def map_from_sampling_content(
+    default_name = to_snake((agent.name or 'PydanticAI Agent').replace(' ', '_'))
+    server_name = server_name or default_name
+    tool_name = tool_name or default_name
+    output_adapter: TypeAdapter[Any] = TypeAdapter(agent.output_type)
```
Fixed - now uses `types_from_output_spec` (mirroring `output_json_schema`) to extract concrete types from the `OutputSpec`, builds a `TypeAdapter` over the single type or a `Union` of multiple types. Added a test exercising `ToolOutput(MyModel)`.

Kludex commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -135,3 +145,56 @@ def map_from_sampling_content(
+
+    async def call_tool(name: str, args: _AgentToolArgs) -> StructuredContent:
+        if name != tool_name:
+            raise ValueError(f'Unknown tool: {name}')
```
Fixed - now uses `{name!r}` per rule:32.

Kludex commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -2,20 +2,30 @@
 
 try:
     from mcp import types as mcp_types
+    from mcp.server.lowlevel.server import Server, StructuredContent
```
Verified - `StructuredContent` is exported by `mcp>=1.25.0` (the locked version), and tests pass. The cached archive Devin inspected was a different version.

Kludex commented on tests/test_mcp.py:
```diff
@@ -2495,3 +2496,154 @@ async def test_server_capabilities_list_changed_fields() -> None:
+# ============================================================================
+
+
+async def _list_tools(server: MCPLowLevelServer) -> mcp_types.ListToolsResult:
+    handler = server.request_handlers[mcp_types.ListToolsRequest]
+    result = await handler(None)
+    assert isinstance(result.root, mcp_types.ListToolsResult)
+    return result.root
+
+
+async def _call_tool(server: MCPLowLevelServer, name: str, args: dict[str, Any]) -> mcp_types.CallToolResult:
+    handler = server.request_handlers[mcp_types.CallToolRequest]
+    req = mcp_types.CallToolRequest(
+        method='tools/call',
+        params=mcp_types.CallToolRequestParams(name=name, arguments=args),
+    )
+    result = await handler(req)
+    assert isinstance(result.root, mcp_types.CallToolResult)
+    return result.root
```
Acknowledged - moving these tests to use a real client/transport is a worthwhile follow-up but out of scope for this PR. The current approach mirrors the pattern used elsewhere in the file and exercises both `list_tools` and `call_tool` (including structured-content serialization for `BaseModel` outputs), so the added handler logic is covered.

Kludex committed:
Add test for union output type in to_mcp

Kludex commented on pydantic_ai_slim/pydantic_ai/agent/abstract.py:
```diff
@@ -1563,8 +1565,50 @@ def to_a2a(
+        """
+        from .._mcp import agent_to_mcp
+
+        warnings.warn('The `to_mcp` method is experimental, and may change in the future.', UserWarning)
```
Why is it experimental?

DouweM replied:
Yeah it shouldn't be

DouweM commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -135,3 +146,61 @@ def map_from_sampling_content(
+    if len(return_types) == 1:
+        output_adapter: TypeAdapter[Any] = TypeAdapter(return_types[0])
+    else:
+        output_adapter = TypeAdapter(Union[tuple(return_types)])  # noqa: UP007
```
Hmm I'd rather have a new method on `agent` for this, like we have `output_json_schema` already

DouweM commented on pydantic_ai_slim/pydantic_ai/agent/abstract.py:
```diff
@@ -1563,8 +1565,50 @@ def to_a2a(
+        server_name: str | None = None,
+        tool_name: str | None = None,
+        tool_description: str | None = None,
+        deps: AgentDepsT = None,
```
Is `deps` enough to send through here, or should we expose other `agent.run` args as well?

DouweM commented on pydantic_ai_slim/pydantic_ai/_mcp.py:
```diff
@@ -135,3 +146,61 @@ def map_from_sampling_content(
+
+        result = await agent.run(user_prompt=args['prompt'], deps=deps)
+
+        return {'result': output_adapter.dump_python(result.output, mode='json')}
```
We should handle `BinaryImage` specifically to turn it into an MCP tool image return.

That may make the output schema wrong though :/

Either way, edge case worth pondering.
